DM-25447: Add support for read-only components

timj opened the pull request:
This was quite a lot more code than I was expecting. There is still an issue over what to do with parameters with read-only components. For now they are sent to the component that is being used to calculate the read-only component and can't be sent to the assembler at all.

timj committed:
Add detector and validPolygon as Exposure components

timj committed:
Add support for StorageClass read-only components

This necessitated a allComponents method to return the
read/write components and read-only components together.

timj committed:
Add some read-only components to Exposure

timj committed:
Use DatasetRef.makeComponentRef simplification API

timj committed:
Remove getStoredItemInfo API

No longer used.

timj committed:
Add a str method for dumping composite map details

timj committed:
Add new specialist formatters for MetricsExample

This allows us to test specialist formatter capabilities
without having to expand the generic YAML and pickle
formatters.  It also allows us to test read parameters
in formatters in addition to assemblers.

timj committed:
Enable true disassembly testing in low-level datastore tests

This enables disassembly testing with the new formatters
without involving butler registry.

timj committed:
Cleanup component datasetType name API usage

timj committed:
Change way parameters are handled if read-only component

We have to process the parameters before we extract the
read-only component.

timj committed:
Add more detail to error message

TallJimbo commented on config/storageClasses.yaml:
```diff
@@ -67,6 +67,8 @@ storageClasses:
   PeakCatalog:
     pytype: lsst.afw.detection.PeakCatalog
+  Polygon:
+    pytype: lsst.afw.geom.Polygon
```
We can also probably save ourselves some work later by associating this storage class with the generic `readFits`/`writeFits` formatter in the datastore config.

TallJimbo commented on python/lsst/daf/butler/core/storageClass.py:
```diff
@@ -165,6 +179,19 @@ def assemblerClass(self) -> Optional[Type]:
         self._assembler = doImport(self._assemblerClassName)
         return self._assembler
 
+    def allComponents(self) -> Dict[str, StorageClass]:
```
If you make this `Mapping` instead of `Dict` now, it'll be easier to change the implementation to, say, `collections.ChainMap` instead of a `dict` copy later.

TallJimbo commented on config/storageClasses.yaml:
```diff
@@ -23,6 +23,14 @@ storageClasses:
+  Point2I:
+    pytype: lsst.geom.Point2I
+  Filter:
+    pytype: lsst.afw.image.Filter
```
This gets me wondering about whether we should allow (maybe just in this context, maybe more broadly) regular Python classes to be used as `StorageClasses` when their definitions are trivial, rather than requiring those to be added explicitly.

timj replied:
That's an interesting thought albeit not for this ticket. So for `StorageClass("lsst.afw.image.Filter")` then rather than defaulting to `object` we assume that the pytype name is the storage class name... (we could even adopt the convention that names that start with upper case are never pytypes). Worth pondering.

TallJimbo commented on tests/test_datastore.py:
```diff
@@ -205,7 +205,71 @@ def testBasicPutGet(self):
+            with self.subTest(storageClass=sc.name):
+                # Create a different dataset type each time round
+                # so that a test failure in this subtest does not trigger
+                # a cascade of tests because of file clashes
```
:100: .  Thanks for thinking about this.

TallJimbo commented on python/lsst/daf/butler/datastores/inMemoryDatastore.py:
```diff
@@ -293,12 +293,27 @@ def get(self, ref: DatasetRef, parameters: Optional[Mapping[str, Any]] = None) -
+        if isReadOnlyComponent:
+            inMemoryDataset = writeStorageClass.assembler().handleParameters(inMemoryDataset, parameters)
+            # Then disable parameters for later
+            parameters = {}
```
Is there a concrete case where this matters?  I'm just wondering if we should say, "you can't pass parameters if you want a read-only component" to try to make things a _little_ simpler.

Edit: after writing the above I saw the commit message that referenced an `origin` parameter to `bbox`.  I'd be okay with just dropping support for that (`origin=PARENT` is all we need - the box returned with `origin=LOCAL` is trivially calculable from the one returned with `origin=PARENT`).  I'd be fine with dropping the `xy0` and `dimensions` read components as being similarly redundant, though I gather that doesn't gain us as much.

timj replied:
I'll ponder this some more when I write up the formatter/assembler documentation. Stating that parameters are never allowed for read-only components is the easiest approach.

TallJimbo commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -385,6 +385,35 @@ def segregateParameters(self, parameters: Optional[Dict[str, Any]] = None) -> Tu
+            from the supplied options.
+        """
+        raise NotImplementedError("This formatter does not support read components")
+
```
Would it make any sense to put this interface on the assembler?  Or maybe make it a fixed mapping in the StorageClass definitions?  I can't think of a case where it would need to depend on the actual formatter used for the components - whether a component can handle a particular `readComponent` seems conceptually like it's a general property of that component's storage class.

Would either or both of those options sidestep the need to write a row with the composite formatter to the datastore records?  The latter at least vaguely sounds like something I think you mentioned as a possibility on Slack, but I don't recall reading why you didn't go with that approach.

timj replied:
You are absolutely right that this look up should be on the assembler and not the formatter. It is a function of the python type so the assembler does know which component is best. This could mean I can remove all the code that added the extra row although I need to ponder how the component can find out the composite storage class so it can get the right assembler. On the other hand the relationship of bbox to Exposure is something that can be worked out without caring about the formatter -- I was using the formatter because the formatter knew about the python type as well but that's clearly superfluous.  I might have to change DatasetType to add a compositeAssembler property.

This is what code review is all about! Granted it means a lot of the code on this ticket is irrelevant.

TallJimbo reviewed (approved):
Code looks good.  I have some of the same concerns on the big picture and increasing complexity that you do, but I'm mostly happy to trust that you've looked into alternatives more than I have and have chosen least-bad options.  On the off chance a thought I had for simplifying things might not be one you've considered, I've left a couple of comments on those big-picture things at various points in the PR.

One more quick big-picture thought: in terms of nomenclature, I wonder if we should not be calling these "readComponents" a type of "component" at all - maybe "properties", with what we're calling "allComponents" here (i.e. real components and properties) -> "attributes"?  Just a thought - I don't like those names enough better that I'd want to push for changing things this late in the game, but the naming right now feels like we started out by thinking of these as another kind of component, and by the end it became clear that they were more different than similar.

timj commented:
I called them read components because from the point of view of the caller they act exactly like components. The caller doesn't even know that some are read/write and some are read-only because they have no idea which of the components ends up being important for disassembly.  We still refer to them as "composite.component" and calling them a property or attribute might be easier to talk about internally (since talking about read-write component relevant for read-only component is annoying at times) but for the general interface they are components.

timj committed:
No longer write component dataset types to registry

We no longer store components at the registry level
so there is not much to be gained by registering
the dataset types.

timj committed:
Do not tell component formatter what component it is

For a disassembled component the formatter is only
interested in reading the entire file, not extracting
a component from the component.

timj committed:
Update queryDatasetTypes now that registry does not include components

timj committed:
Add makeAllComponentDatasetTypes method

This creates dataset types for all the components.

timj committed:
Record parent storage class in composite dataset types

timj committed:
Use modern DatasetType API in tests

timj committed:
Remove duplication of test formatter definitions between posix and s3

timj committed:
Enable read-only component reading from disassembled components

This requires a new method on assembler to return which component
should be used to calculate the read-only component.

timj committed:
Move where parameter validation occurs

Parameters for read-only components are problematic since it
is not entirely clear whether the parameters should be applied
to the component that is used to calculate the read-only
component, or should be applied to the calculation of
the read-only component itself.

The complication is that assemblers must support the same
parameters as formatters (otherwise in-memory datastore
can not function) and also can apply parameters that
the formatter did not understand. In the current implementation
this means that the assembler can only see the final storage
class and at that point parameters might not be relevant.

As concrete examples. In the test suite I have added
a "counter" read-only component that returns the
number of elements in metrics.data.  This means that
the assembler sees an Integer storage class and integers
aren't amenable to parameters. Instead the "slice" parameter
is assumed to apply to the "data" component and then
"counter" is calculated on the sliced "data".

Another example is bbox in Exposure. This can be applied
to the "image" component. Where should parameters go?
To the calculation of the "image" or to the calculation
of the bounding box (origin is the only one). Once
the bounding box is created though the "origin" parameter
has no meaning so "origin" is only relevant when passed
to the "image" and can't be relevant to the assembler.

Therefore read-only components for disassembled
composites do not receive any parameters.

timj committed:
Add Polygon formatter definition

TallJimbo commented on python/lsst/daf/butler/registry/datasets/byDimensions/_manager.py:
```diff
@@ -141,7 +149,7 @@ def register(self, datasetType: DatasetType) -> Tuple[DatasetRecordStorage, bool
         if inserted and datasetType.isComposite:
             for component in datasetType.storageClass.allComponents():
-                self.register(datasetType.makeComponentDatasetType(component))
+                pass  # self.register(datasetType.makeComponentDatasetType(component))
```
Looks like you can now remove this whole block.

TallJimbo replied:
And should we also add in this function to make sure no one tries to register just a component?  There are probably other places we could put that kind of check, but this is the only one I can think of where it might actually succeed (and mess up the registry) rather than failing with some possibly-less-friendly error.

timj replied:
Yes. I did mean to remove that. I can add a check here.

TallJimbo commented on tests/test_butler.py:
```diff
@@ -686,14 +686,18 @@ def testGetDatasetTypes(self):
 
         # When a DatasetType is added to the registry entries are not created
-        # for components.
+        # for components but querying them can reeturn the components.
```
typo: reeturn

TallJimbo commented on python/lsst/daf/butler/core/datasets/type.py:
```diff
@@ -273,6 +274,21 @@ def makeComponentDatasetType(self, component: str) -> DatasetType:
+            then returns an empty list.
+        """
+        if not self.isComposite():
+            return []
```
Do you actually need this check?  I'd have thought `storageClass.allComponents()` would return an empty iterable in this case anyway.

timj replied:
Yes. I was prematurely optimizing.

TallJimbo commented on python/lsst/daf/butler/core/datasets/type.py:
```diff
@@ -81,12 +84,17 @@ class DatasetType:
     storageClass : `StorageClass` or `str`
         Instance of a `StorageClass` or name of `StorageClass` that defines
         how this `DatasetType` is persisted.
+    parentStorageClass : `StorageClass` or `str`
```
Should probably be `, optional`.

TallJimbo commented on python/lsst/daf/butler/core/datasets/type.py:
```diff
@@ -131,6 +140,28 @@ def __init__(self, name: str, dimensions: Union[DimensionGraph, Iterable[Dimensi
+        _, componentName = self.splitDatasetTypeName(self._name)
+        if parentStorageClass is None and componentName is not None:
+            log.warning("Component dataset type '%s' constructed without parent storage class",
+                        self._name)
```
Unless this fails tests that are a pain to fix, I'd be happy to bump this up to an exception.

timj replied:
I set it to be an exception to find where tests were breaking. I think a couple of tests were still morning (mostly the ones that set up fake datasetrefs). -- I'll take another look at them. Raising an exception will mean that this code will fail if an old registry is used when the code encounters the component dataset types but I imagine that's probably okay.

TallJimbo commented on python/lsst/daf/butler/core/datasets/type.py:
```diff
@@ -179,6 +210,20 @@ def storageClass(self) -> StorageClass:
+        with this  `DatasetType` is persisted. Note that if DatasetType was
+        constructed with a name of a StorageClass then Butler has to be
+        initialized before using this property. Can be `None` if this is
+        not a component of a composite.
```
Docstring should be split into brief + extended descriptions.

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -899,8 +899,29 @@ def get(self, ref: DatasetRef, parameters: Optional[Mapping[str, Any]] = None) -
+            compositeStorageClass = ref.datasetType.parentStorageClass
+
+            # Assume something else will handle things if we can not find
+            # it
```
Is this actually a reasonable assumption?  Is there some other part of the system that could handle this case?

timj replied:
If the parent storage class is not available this will assume it's a real component being requested. At that point the branch below that looks up the disassembled component will fail and you'll get an error about the component not being recognized.

TallJimbo reviewed (approved):
Looks good!  I'm glad the last review pass was useful.  I don't have much to add in this pass - just minor comments - though some of the old ones are still relevant as well.

timj committed:
Raise exception if DatasetType has parent storage class but is not component

Check that a component does have a parent but a composite does not.
The parent storage class is not checked to determine if the component
is defined by it.

timj committed:
Add some tests for DatasetType parentStorageClass

This required changes to how pickling and deep copy worked.
Also to simplify pickling the parentStorageClass is now
also a positional argument.

timj committed:
Add special placeholder parent storage class and update method

This allows pipeline definitions to set a temporary parent
when analyzing the pipeline but then during execution update
it with the real parent.
